blockchain/utxocache.go

```diff
@@ -0,0 +1,793 @@
+
+	return baseEntrySize + uint64(len(entry.pkScript))
+}
+
+// Spend marks the output as spent.  Spending an output that is already spent
+// has no effect.
+func (entry *UtxoEntry) Spend() {
+	// Nothing to do if the output is already spent.
+	if entry.IsSpent() {
+		return
+	}
+
+	// Mark the output as spent and modified.
+	entry.packedFlags |= tfSpent | tfModified
+}
+
+// Clone returns a shallow copy of the utxo entry.
+func (entry *UtxoEntry) Clone() *UtxoEntry {
+	if entry == nil {
+		return nil
+	}
+
+	return &UtxoEntry{
+		amount:      entry.amount,
+		pkScript:    entry.pkScript,
+		blockHeight: entry.blockHeight,
+		packedFlags: entry.packedFlags,
+	}
+}
+
+// utxoView is a common interface for structures that implement a UTXO view.
+type utxoView interface {
+	// getEntry tries to get an entry from the view.  If the entry is not in the
+	// view, both the returned entry and the error are nil.
+	getEntry(outpoint wire.OutPoint) (*UtxoEntry, error)
+
+	// addEntry adds a new entry to the view.  Set overwrite to true if this
+	// entry should overwrite any existing entry for the same outpoint.
+	addEntry(outpoint wire.OutPoint, entry *UtxoEntry, overwrite bool) error
+	// spendEntry marks an entry as spent.
+	spendEntry(outpoint wire.OutPoint, entry *UtxoEntry) error
+}
+
+// utxoByHashSource is an interface that allows fetching UTXO entries by
+// transaction hash.
+// This interface exists due to the legacy spend journal database structure.
+type utxoByHashSource interface {
+	// getEntryByHash looks for an entry with the given transaction hash.
+	// This method exists due to the legacy spend journal database structure.
+	// Its execution is very inefficient, but it's almost never used.
+	getEntryByHash(hash *chainhash.Hash) (*UtxoEntry, error)
+}
+
+// utxoCache is a cached utxo view in the chainstate of a BlockChain.
+//
+// It implements the utxoView interface, but should only be used as such with the
+// state mutex held.  It also implements the utxoByHashSource interface.
+type utxoCache struct {
+	db database.DB
+
+	// maxTotalMemoryUsage is the maximum memory usage in bytes that the state
+	// should contain in normal circumstances.
+	maxTotalMemoryUsage uint64
+
+	// This mutex protects the internal state.
+	// A simple mutex instead of a read-write mutex is chosen because the main
+	// read method also possibly does a write on a cache miss.
+	mtx sync.Mutex
+
+	// cachedEntries keeps the internal cache of the utxo state.  The tfModified
+	// flag indicates that the state of the entry (potentially) deviates from the
+	// state in the database.  Explicit nil values in the map are used to
+	// indicate that the database does not contain the entry.
+	cachedEntries    map[wire.OutPoint]*UtxoEntry
+	totalEntryMemory uint64 // Total memory usage in bytes.
+	lastFlushHash    chainhash.Hash
+}
+
+// newUtxoState initiates a new utxo state instance.
+func newUtxoState(db database.DB, maxTotalMemoryUsage uint64) *utxoCache {
```

Comment:
godoc needs update?

Reply:
True that. Had a totally different implementation at first that was called utxoState :)